File: statsig-php/src/EvaluationTypes/Layer.php
<?php

namespace Statsig\EvaluationTypes;

use Statsig\StatsigFFI;

class Layer extends BaseEvaluation
{
    public readonly ?string $groupName;
    public readonly ?string $allocatedExperimentName;
    private readonly array $__value;
    private readonly string $__statsig_ref;

    public function __construct(string $raw_result, string $__statsig_ref)
    {
        $result = json_decode($raw_result, true);
        parent::__construct($raw_result, $result);
        $this->__statsig_ref = $__statsig_ref;
        $this->__value = $result['__value'] ?? [];
        $this->groupName = $result['group_name'] ?? null;
        $this->allocatedExperimentName = $result['allocated_experiment_name'] ?? null;
    }

    public function get(string $param_name, $fallback)
    {
        return $this->getValueImpl(
            $this->__value,
            $param_name,
            $fallback,
            function ($param_name) {
                $this->logParameterExposure($param_name);
            }
        );
    }

    private function logParameterExposure(string $param_name): void
    {
        StatsigFFI::get()->statsig_log_layer_param_exposure($this->__statsig_ref, $this->__raw_result, $param_name);
    }
}

File: statsig-go/src/layer.go
package statsig

/*
#cgo CFLAGS: -I../../statsig-ffi/include
#include "statsig_ffi.h"

#include <stdlib.h>
*/
import (
	"C"
)
import "github.com/statsig-io/statsig-server-core/statsig-go/src/utils"

type Layer struct {
	Name                    string                 `json:"name"`
	Value                   map[string]interface{} `json:"__value"`
	RuleID                  string                 `json:"rule_id"`
	EvaluationDetails       EvaluationDetails      `json:"details"`
	IdType                  string                 `json:"id_type"`
	GroupName               string                 `json:"group_name"`
	AllocatedExperimentName string                 `json:"allocated_experiment_name"`
	RawJSON                 string                 `json:"-"`
	statsigInstance         Statsig
	disableExposureLogging  bool
}

func (l *Layer) setStatsigInstance(statsig *Statsig) {
	l.statsigInstance = *statsig
}

func (l *Layer) setDisableExposureLogging(disableExposureLogging bool) {
	l.disableExposureLogging = disableExposureLogging
}

func (l *Layer) setRawResult(json string) {
	l.RawJSON = json
}

func (l *Layer) Get(u *StatsigUser, paramName string) interface{} {
	return l.getValueImpl(paramName, u)
}

func (l *Layer) getValueImpl(paramName string, u *StatsigUser) interface{} {
	val, ok := l.Value[paramName]

	if !ok {
		return ""
	}

	if !l.disableExposureLogging {
		l.logLayerParamExposure(paramName)
	}
	return val
}

func (l *Layer) logLayerParamExposure(paramName string) {
	C.statsig_log_layer_param_exposure(
		C.uint64_t(l.statsigInstance.InnerRef),
		C.CString(l.RawJSON),
		C.CString(paramName),
	)
}
func (l *Layer) GetString(key string, fallback string) string {
	return utils.GetTypedValue(l.Value, key, fallback)
}

func (l *Layer) GetNumber(key string, fallback float64) float64 {
	return utils.GetTypedValue(l.Value, key, fallback)
}

func (l *Layer) GetBool(key string, fallback bool) bool {
	return utils.GetTypedValue(l.Value, key, fallback)
}

func (l *Layer) GetSlice(key string, fallback []interface{}) []interface{} {
	return utils.GetTypedValue(l.Value, key, fallback)
}

func (l *Layer) GetMap(key string, fallback map[string]interface{}) map[string]interface{} {
	return utils.GetTypedValue(l.Value, key, fallback)
}
